test/infrastructure/docker/exp/docker/nodepool.go

```diff
@@ -83,67 +83,84 @@ func NewNodePool(kClient client.Client, cluster *clusterv1.Cluster, mp *clusterv
 	return np, nil
 }
 
-// ReconcileMachines will build enough machines to satisfy the machine pool / docker machine pool spec and update the
-// docker machine pool status
-func (np *NodePool) ReconcileMachines(ctx context.Context) error {
-	matchingMachineCount := int32(len(np.machinesMatchingInfrastructureSpec()))
-	if matchingMachineCount < *np.machinePool.Spec.Replicas {
-		for i := int32(0); i < *np.machinePool.Spec.Replicas-matchingMachineCount; i++ {
-			if err := np.addMachine(ctx); err != nil {
-				return errors.Wrap(err, "failed to create a new docker machine")
+// ReconcileMachines will build enough machines to satisfy the machine pool / docker machine pool spec
+// eventually delete all the machine in excess, and update the status for all the machines.
+//
+// NOTE: The goal for the current implementation is to verify MachinePool construct; accordingly,
+// currently the nodepool supports only a recreate strategy for replacing old nodes with new ones
+// (all existing machines are killed before new ones are created).
+// TODO: consider if to support a Rollout strategy (a more progressive node replacement).
+func (np *NodePool) ReconcileMachines(ctx context.Context, log logr.Logger) error {
+	desiredReplicas := int(*np.machinePool.Spec.Replicas)
+
+	// Delete all the machines in excess (outdated machines or machines exceeding desired replica count).
+	machineDeleted := false
+	totalNumberOfMachines := 0
+	for _, machine := range np.machines {
+		if totalNumberOfMachines+1 > desiredReplicas || !np.isMachineMatchingInfrastructureSpec(machine) {
```

Comment:
By moving the increment of `totalNumberOfMachines` above this line, you can avoid doing +1

Reply:
This requires to decrement in case a machine gets delete, but done